src/reason-parser/reason_pprint_ast.ml

```diff
@@ -6640,6 +6644,159 @@ class printer  ()= object(self:'self)
     match xs with
     | [x] -> x
     | xs -> makeBreakableList xs
+
+  method formatFunAppl ~jsxAttrs ~args ~funExpr () =
+    (* If there was a JSX attribute BUT JSX component wasn't detected,
+       that JSX attribute needs to be pretty printed so it doesn't get
+       lost *)
+    let maybeJSXAttr = List.map self#attribute jsxAttrs in
+
+    let categorizeFunApplArgs args =
+      let reverseArgs = List.rev args in
+      match reverseArgs with
+      | (Nolabel, {pexp_desc = Pexp_fun _})::_ -> `LastArgIsCallback reverseArgs
+      | _ -> `NormalFunAppl args
+    in
+    begin match categorizeFunApplArgs args with
```

Comment:
This printer is getting pretty long. Can we move this to another module? Something like "style_classifier.ml". We can probably move a bunch of stuff there. It would make Merlin more responsive while editing the pretty printer.